Documentation for the scheduling replicator

nickva opened the pull request:
"What's New" update includes feature improvements and also an implicit to 3.0

"Replicator" section describing some new behavior and includes a state transition diagram.

Jira: COUCHDB-3324

tonysun83 commented on src/whatsnew/3.0.rst:
```diff
@@ -0,0 +1,105 @@
+=============
+
+* New scheduling replicator. The core of the new replicator is a
+  scheduler whichallows running a large number of replication
```
whichallows needs a whitespace

tonysun83 commented on src/replication/replicator.rst:
```diff
@@ -155,98 +203,192 @@ the following order:
-you can for example define a view which maps replication IDs to document IDs.
+Notice the state for this replication is ``failed``. Unlike
+``crashing``, ``failed`` state is terminal. As long as both documents
+are present replicator will not retry to run this replication. Another
```
This comment was a little ambiguous: "As long as both documents are present, replicator will not retry to run this replication." So if you have RepA and RepB that describe the same replication, initially RepB will fail as it describes the same replication and RepA is already running, so RepB will be in "failed" state. Let's say something occurs, and RepA needs to crash. So based on the comment, it would seem RepA and RepB both won't start up. I would assume RepA would start up again right?

nickva replied:
RepA crashing won't start RepB automatically. They both specify the same replication so if RepA crashes RebB will also crash probably. I'll try to reword it better a bit.

tonysun83 commented on src/replication/replicator.rst:
```diff
@@ -155,98 +203,192 @@ the following order:
+Notice the state for this replication is ``failed``. Unlike
+``crashing``, ``failed`` state is terminal. As long as both documents
+are present replicator will not retry to run this replication. Another
+reason for could be malformed documents. For example if worker process
```
reason for _crashes_ could be?

tonysun83 commented on src/replication/replicator.rst:
```diff
@@ -155,98 +203,192 @@ the following order:
+way of creating replications, but posting to the ``_replicate`` HTTP
+endpoint is also supported.
+
+.. note:: Replications created through the ``_replicate`` endpoint
```
Also, I'm not sure if we addressed this, but with _replicate, if you post an "_id", the replication, it doesn't persist:
https://issues.apache.org/jira/browse/COUCHDB-3235

nickva replied:
Oh good one. We should fix that. But let's fix it after the merge so we don't have to fix it twice. We'll probably just strip _id from the _replicate before it gets processed

wohali replied:
As we're talking about getting a 2.1.0 out the door, let's raise the priority on COUCHDB-3235 and get that fix in before we release.

tonysun83 commented on src/replication/replicator.rst:
```diff
@@ -155,98 +203,192 @@ the following order:
+
+ * ``Pending``: Replication job is not running and is waiting its
+   turn. This state is reached then more than ``replicator.max_jobs``
+   replication have been added to the scheduler. In that case
```
This state is reached when replication jobs added to the scheduler exceeds replicator.maxjobs.

tonysun83 commented on src/replication/replicator.rst:
```diff
@@ -155,98 +203,192 @@ the following order:
+running replication, and re-evaluate their cluster placement in light
+of the new set of live nodes.  This mechanism also provides
+replication fail-over in case on node fails. Replication jobs started
+form replication documents (but not those started from a ``_replicate``
```
from

tonysun83 commented:
Nice documentation! Very informative, especially with the diagram.

nickva commented:
@tonysun83 Thanks for taking  a look

nickva commented:
@sagelywizard Good ideas!

wohali commented:
Since we're talking about moving this into 2.1.0 now, please `mv src/whatsnew/3.0.rst src/whatsnew/2.1.rst` ?

wohali commented on src/config/replicator.rst:
```diff
@@ -23,15 +23,51 @@ Replicator Database Configuration
+    .. config:option:: max_jobs
 
-    .. config:option:: max_replication_retry_count
+        .. versionadded:: 3.0
```
Also here, might be 2.1 instead of 3.0, and for all such instances in this document. Pending discussion about which version should be the target in IRC/mailing lists :)

nickva replied:
Fixed

wohali commented on src/replication/replicator.rst:
```diff
@@ -16,6 +16,11 @@
 Replicator Database
 ===================
 
+.. versionchanged:: 3.0.0 Scheduling replicator was
```
Again version number here subject to change.

nickva replied:
2.1.0 as was agreed in the IRC channel

wohali commented on src/replication/replicator.rst:
```diff
@@ -43,107 +48,151 @@ Let's say you POST the following document into ``_replicator``:
     {
         "_id": "my_rep",
-        "source":  "http://myserver.com:5984/foo",
+        "source":  "http://user:pass@localhost:5984/foo",
```
Perhaps a different hostname here, just to make sure it's still clear that replication is not a "private to me" concept, but a "works between machines" concept?

nickva replied:
Good idea

wohali commented on src/replication/replicator.rst:
```diff
@@ -43,107 +48,151 @@ Let's say you POST the following document into ``_replicator``:
+        socket options are: [{keepalive,true},{nodelay,false}]
+    [notice] 2017-04-05T17:16:19.646759Z node1@127.0.0.1 <0.29432.0> -------- Document `my_rep` triggered replication `a81a78e822837e66df423d54279c15fe+continuous+create_target`
+
+Replication state of this document can then be queried from
```
Do you want to mention here about what happens to the `_replicator` document as well? States are not written back, but there is still a document lifecycle, correct?

nickva replied:
Agreed

nickva replied:
I am thinking of adding this

```
Replication document contents stay the same. Previously, before version
2.1, it was updated with the ``triggered`` state.
```

Above in `versionchanged` section there is mention of it as well, then we mention how the whole thing as well ```documents
added to it will trigger replications. Create (``PUT`` or ``POST``) a document
to start replication. ``DELETE`` a replication document to cancel an ongoing
replication.```

wohali commented on src/replication/replicator.rst:
```diff
@@ -154,98 +203,253 @@ the following order:
-        "source":  "http://myserver.com:5984/foo",
-        "target":  "http://user:pass@localhost:5984/bar"
+        "_id": "my_rep",
+        "source":  "http://user:pass@localhost:5984/foo",
```
Same comment as before, would be happier to see source and target not being on the same host.

wohali commented on src/replication/replicator.rst:
```diff
@@ -154,98 +203,253 @@ the following order:
-        "_id": "doc_B",
-        "source":  "http://myserver.com:5984/foo",
+        "_id": "my_rep_dup",
+        "source":  "http://user:pass@localhost:5984/foo",
```
And here.

wohali commented on src/replication/replicator.rst:
````diff
@@ -154,98 +203,253 @@ the following order:
-    }
+Both describe exactly the same replication (only their ``_ids``
+differ). In this case document ``my_rep`` triggers the
+replication. While ``my_rep_dup``` will fail. Inspecting
````
Sentence fragment, please revise. This should read `replication, while`

wohali commented on src/replication/replicator.rst:
```diff
@@ -154,98 +203,253 @@ the following order:
-you can for example define a view which maps replication IDs to document IDs.
+Notice the state for this replication is ``failed``. Unlike
+``crashing``, ``failed`` state is terminal. As long as both documents
+are present replicator will not retry to run my_rep_dub
```
"...are present *the* replicator will..."

wohali replied:
replace "my_rep_dub" with ``my_rep_dup`` (spelling and typeface bug)

wohali commented on src/replication/replicator.rst:
```diff
@@ -154,98 +203,253 @@ the following order:
+are present replicator will not retry to run my_rep_dub
+replication. Another reason could be malformed documents. For
+example if worker process count is specified as a string
+(``"worker_processes": "a few"``) instead of an integer.
```
Sentence fragment: "...instead of an integer, failure will occur."

wohali commented on src/replication/replicator.rst:
```diff
@@ -154,98 +203,253 @@ the following order:
+Once replication jobs are created they are managed by the scheduler.
+The scheduler is the replication component which periodically stops
+some jobs and starts others. This behavior makes it posssible to have a
+larger number of jobs than what the cluster could run at one time.
```
awkward, try "...larger number of jobs than the cluster is capable of running simultaneously."

nickva replied:
Much better, thank you!

wohali commented on src/replication/replicator.rst:
```diff
@@ -154,98 +203,253 @@ the following order:
+
+     Replication state diagram
+
+Green and red shapes represent replication job states.
```
Recent figures suggest that approximately 10% of men have red-green colour blindness. Those people won't be able to tell the difference between your healthy and unhealthy states, especially since they differ only in colour.

Can we change the colours used from green/red/yellow to an alternative palette? Alternatively, might it make sense for unhealthy states to have a slightly different shape?

Here are 2 excellent resources I've used in the past to help with this problem:
* http://www.visibone.com/colorblind/
* http://safecolours.rigdenage.com/

nickva replied:
Blue/Yellow might work I think

wohali commented on src/replication/replicator.rst:
```diff
@@ -154,98 +203,253 @@ the following order:
+   once completed they will be removed from the system, so their
+   completion state is preserved in ``_scheduler/jobs`` endpoint.
+
+White shapes indicate internal API boundaries and point to how the
```
Describe the shapes by # of sides as well as colour here if you could, please.

wohali commented on src/replication/replicator.rst:
```diff
@@ -154,98 +203,253 @@ the following order:
+Before explaining the details of each state, it is worth noticing that
+color and shape of each state in the diagram:
+
+`Green` vs `red` partitions states into "healthy" and "unhealthy",
```
Same comment as previous regarding red/green colour blindness.

wohali commented on src/replication/replicator.rst:
```diff
@@ -154,98 +203,253 @@ the following order:
+ * ``Failed``: Replication document could not be processed and turned
+   into a valid replication job for the scheduler. This state is
+   terminal and requires user intervention to fix the problem. Typical
+   reasons for ending up in this state is a malformed document. For
```
Invalid subject-verb agreement: "reasons" vs "is".

wohali commented on src/replication/replicator.rst:
```diff
@@ -154,98 +203,253 @@ the following order:
+   terminal and requires user intervention to fix the problem. Typical
+   reasons for ending up in this state is a malformed document. For
+   example specifying an integer for a parameter which accepts a
+   boolean. Another reason could be specifying a duplicate
```
Sentence fragments, revise. "For example, specifying an integer for a parameter which accepts a boolean will result in this state. Another reason for failure could be..."

wohali commented on src/replication/replicator.rst:
```diff
@@ -154,98 +203,253 @@ the following order:
+   would be processed and posted to the target. Job is still
+   considered ``Running`` even if its workers are currently not
+   streaming changes from source to target and are just waiting on the
+   change feed.
```
You might want to note that this would be a normal state for a continuous replication explicitly.

wohali commented on src/replication/replicator.rst:
```diff
@@ -154,98 +203,253 @@ the following order:
+   database, a permissions error, etc. Repeated consecutive crashes
+   result in an exponential backoff. This state is considered temporary
+   (non-terminal) and replication jobs will be periodically
+   retried. Maximum backoff interval is around a day or so.
```
Is there any way to retrieve the current backoff timer?

nickva replied:
Not as easily. We don't explicitly schedule a job to run in the future, say something like "this job shall run again at 5pm later today" instead every time we reschedule we look in the history of each job and see how many consecutive crashes it has experienced.

wohali commented on src/replication/replicator.rst:
```diff
@@ -154,98 +203,253 @@ the following order:
+
+Previous version of CouchDB replicator wrote state updates back to
+replication documents. In case were user code programmatically read
+those states, there is compatibility mode enabled via a configuration
```
"In cases where user code..."

wohali commented on src/replication/replicator.rst:
```diff
@@ -154,98 +203,253 @@ the following order:
+the replication. To update a replication, for example, change the
+number of worker or the source, simply update the document with new
+data. If there is extra application-specific data in the replication
+documents, that data is ignored by the replicator.
```
Is there a way to cancel a `_replicate` triggered replication? If so I'd imagine it's via the new endpoint, and you probably want to explicitly show an example. Otherwise note that `_replicate` triggered replications cannot be cancelled.

nickva replied:
_replicate replications work as before. They are cancelled by either POST-ing to _replicate with an `{"id":<repid>` and `"cancel":true}` or specifying the same replication parameters as used during creation with the addition of `"cancel":true` field.

wohali commented on src/replication/replicator.rst:
```diff
@@ -154,98 +203,253 @@ the following order:
+application will notice the change, rescan all the document and
+running replication, and re-evaluate their cluster placement in light
+of the new set of live nodes.  This mechanism also provides
+replication fail-over in case on node fails. Replication jobs started
```
"...in case one node fails" or "...in case a node fails."

wohali commented on src/replication/replicator.rst:
```diff
@@ -154,98 +203,253 @@ the following order:
+running replication, and re-evaluate their cluster placement in light
+of the new set of live nodes.  This mechanism also provides
+replication fail-over in case on node fails. Replication jobs started
+form replication documents (but not those started from ``_replicate``
```
"from" not "form"

wohali commented:
Where's the updates to the API Reference, especially for the new scheduler endpoints? I'm especially curious if any of them support other methods/verbs than GET.

nickva commented:
@wohali Added some improvement as requested:

 * API sections for _scheduler/jobs and _scheduler/docs
 * Replaced source with a non-localhost server URL to emphasize replication across machines
 * Fixed embarrassing grammar mistakes
 * Good call on green/red color scheme, regarding color blindness. Now it's blue/yellow.

wohali commented:
+1, go for it

nickva commented:
Thank you!
